File: src/parser/parser.rs
use crate::parser::ast::*;
use crate::parser::lexer::Token;

pub struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Parser { tokens, pos: 0 }
    }

    fn peek(&self) -> &Token {
        self.tokens.get(self.pos).unwrap_or(&Token::Eof)
    }

    /// Consume and return the current token.
    fn advance(&mut self) -> Token {
        let tok = self.tokens.get(self.pos).cloned().unwrap_or(Token::Eof);
        if self.pos < self.tokens.len() {
            self.pos += 1;
        }
        tok
    }

    fn expect(&mut self, expected: &Token) -> Result<(), String> {
        if self.peek() == expected {
            self.advance();
            Ok(())
        } else {
            Err(format!("expected {:?}, got {:?}", expected, self.peek()))
        }
    }

    // ── Top-level entry point ────────────────────────────────────────────────

    pub fn parse(&mut self) -> Result<Statement, String> {
        match self.peek() {
            Token::Select => self.parse_select(),
            _ => {
                let tok = self.peek().clone();
                Err(format!("expected SELECT, got {:?}", tok))
            }
        }
    }

    // ── SELECT statement ─────────────────────────────────────────────────────

    fn parse_select(&mut self) -> Result<Statement, String> {
        self.advance(); // consume SELECT

        let columns = self.parse_select_list()?;

        self.expect(&Token::From)?;
        let from = self.parse_table_ref()?;

        // Zero or more JOINs
        let mut joins = Vec::new();
        while matches!(
            self.peek(),
            Token::Join | Token::Inner | Token::Left | Token::Right | Token::Cross
        ) {
            joins.push(self.parse_join()?);
        }

        let where_clause = if self.peek() == &Token::Where {
            self.advance();
            Some(self.parse_expr()?)
        } else {
            None
        };

        let group_by = if self.peek() == &Token::Group {
            self.advance();
            self.expect(&Token::By)?;
            self.parse_expr_list()?
        } else {
            Vec::new()
        };

        let having = if self.peek() == &Token::Having {
            self.advance();
            Some(self.parse_expr()?)
        } else {
            None
        };

        let order_by = if self.peek() == &Token::Order {
            self.advance();
            self.expect(&Token::By)?;
            self.parse_order_by_list()?
        } else {
            Vec::new()
        };

        let limit = if self.peek() == &Token::Limit {
            self.advance();
            match self.advance() {
                Token::Integer(n) => Some(n as usize),
                tok => return Err(format!("expected integer after LIMIT, got {:?}", tok)),
            }
        } else {
            None
        };

        Ok(Statement::Select(SelectStatement {
            columns,
            from,
            joins,
            where_clause,
            group_by,
            having,
            order_by,
            limit,
        }))
    }

    // ── SELECT list ──────────────────────────────────────────────────────────

    fn parse_select_list(&mut self) -> Result<Vec<SelectItem>, String> {
        let mut items = Vec::new();
        loop {
            if self.peek() == &Token::Star {
                // Bare `*`
                self.advance();
                items.push(SelectItem::Wildcard);
            } else {
                let expr = self.parse_expr()?;

                // `table.*` is emitted by parse_primary as Column { table: Some(t), column: "*" }.
                // Convert it to QualifiedWildcard here in the SELECT-list context.
                let item = match expr {
                    Expression::Column { table: Some(t), column } if column == "*" => {
                        SelectItem::QualifiedWildcard(t)
                    }
                    other => {
                        let alias = if self.peek() == &Token::As {
                            self.advance();
                            match self.advance() {
                                Token::Identifier(name) => Some(name),
                                tok => {
                                    return Err(format!(
                                        "expected alias name after AS, got {:?}",
                                        tok
                                    ))
                                }
                            }
                        } else if let Token::Identifier(_) = self.peek() {
                            // Implicit alias: `expr name`
                            match self.advance() {
                                Token::Identifier(name) => Some(name),
                                _ => unreachable!(),
                            }
                        } else {
                            None
                        };
                        SelectItem::Expression(other, alias)
                    }
                };
                items.push(item);
            }

            if self.peek() == &Token::Comma {
                self.advance();
            } else {
                break;
            }
        }
        Ok(items)
    }

    // ── Table reference ──────────────────────────────────────────────────────

    fn parse_table_ref(&mut self) -> Result<TableRef, String> {
        let table_name = match self.advance() {
            Token::Identifier(name) => name,
            tok => return Err(format!("expected table name, got {:?}", tok)),
        };

        let alias = if self.peek() == &Token::As {
            self.advance();
            match self.advance() {
                Token::Identifier(name) => Some(name),
                tok => return Err(format!("expected alias after AS, got {:?}", tok)),
            }
        } else if let Token::Identifier(_) = self.peek() {
            match self.advance() {
                Token::Identifier(name) => Some(name),
                _ => unreachable!(),
            }
        } else {
            None
        };

        Ok(TableRef { table_name, alias })
    }

    // ── JOIN clause ──────────────────────────────────────────────────────────

    fn parse_join(&mut self) -> Result<Join, String> {
        let join_type = match self.peek().clone() {
            Token::Join => {
                self.advance();
                JoinType::Inner
            }
            Token::Inner => {
                self.advance();
                self.expect(&Token::Join)?;
                JoinType::Inner
            }
            Token::Left => {
                self.advance();
                self.expect(&Token::Join)?;
                JoinType::Left
            }
            Token::Right => {
                self.advance();
                self.expect(&Token::Join)?;
                JoinType::Right
            }
            Token::Cross => {
                self.advance();
                self.expect(&Token::Join)?;
                JoinType::Cross
            }
            tok => return Err(format!("expected JOIN keyword, got {:?}", tok)),
        };

        let table = self.parse_table_ref()?;
        self.expect(&Token::On)?;
        let condition = self.parse_expr()?;

        Ok(Join { join_type, table, condition })
    }

    // ── ORDER BY list ────────────────────────────────────────────────────────

    fn parse_order_by_list(&mut self) -> Result<Vec<OrderByItem>, String> {
        let mut items = Vec::new();
        loop {
            let expression = self.parse_expr()?;
            let ascending = match self.peek() {
                Token::Asc => {
                    self.advance();
                    true
                }
                Token::Desc => {
                    self.advance();
                    false
                }
                _ => true,
            };
            items.push(OrderByItem { expression, ascending });

            if self.peek() == &Token::Comma {
                self.advance();
            } else {
                break;
            }
        }
        Ok(items)
    }

    // ── Comma-separated expression list ──────────────────────────────────────

    fn parse_expr_list(&mut self) -> Result<Vec<Expression>, String> {
        let mut exprs = Vec::new();
        loop {
            exprs.push(self.parse_expr()?);
            if self.peek() == &Token::Comma {
                self.advance();
            } else {
                break;
            }
        }
        Ok(exprs)
    }

    // ── Expression parsing (recursive descent by precedence) ─────────────────
    //
    // Precedence levels (lowest → highest):
    //   OR
    //   AND
    //   NOT  (prefix, right-associative)
    //   comparison  (=  <>  <  >  <=  >=)   — non-chaining
    //   additive    (+  -)
    //   multiplicative  (*  /)
    //   unary minus  (-)                     — right-associative
    //   primary      (literal, column, function, parenthesised expr)

    pub fn parse_expr(&mut self) -> Result<Expression, String> {
        self.parse_or()
    }

    fn parse_or(&mut self) -> Result<Expression, String> {
        let mut left = self.parse_and()?;
        while self.peek() == &Token::Or {
            self.advance();
            let right = self.parse_and()?;
            left = Expression::BinaryOp {
                left: Box::new(left),
                op: BinaryOperator::Or,
                right: Box::new(right),
            };
        }
        Ok(left)
    }

    fn parse_and(&mut self) -> Result<Expression, String> {
        let mut left = self.parse_not()?;
        while self.peek() == &Token::And {
            self.advance();
            let right = self.parse_not()?;
            left = Expression::BinaryOp {
                left: Box::new(left),
                op: BinaryOperator::And,
                right: Box::new(right),
            };
        }
        Ok(left)
    }

    fn parse_not(&mut self) -> Result<Expression, String> {
        if self.peek() == &Token::Not {
            self.advance();
            let expr = self.parse_not()?; // right-associative: NOT NOT a => Not(Not(a))
            Ok(Expression::UnaryOp {
                op: UnaryOperator::Not,
                expr: Box::new(expr),
            })
        } else {
            self.parse_comparison()
        }
    }

    fn parse_comparison(&mut self) -> Result<Expression, String> {
        let left = self.parse_additive()?;
        let op = match self.peek() {
            Token::Eq  => BinaryOperator::Eq,
            Token::Neq => BinaryOperator::Neq,
            Token::Lt  => BinaryOperator::Lt,
            Token::Gt  => BinaryOperator::Gt,
            Token::Lte => BinaryOperator::Lte,
            Token::Gte => BinaryOperator::Gte,
            _ => return Ok(left),
        };
        self.advance();
        let right = self.parse_additive()?;
        Ok(Expression::BinaryOp {
            left: Box::new(left),
            op,
            right: Box::new(right),
        })
    }

    fn parse_additive(&mut self) -> Result<Expression, String> {
        let mut left = self.parse_multiplicative()?;
        loop {
            let op = match self.peek() {
                Token::Plus  => BinaryOperator::Add,
                Token::Minus => BinaryOperator::Sub,
                _ => break,
            };
            self.advance();
            let right = self.parse_multiplicative()?;
            left = Expression::BinaryOp {
                left: Box::new(left),
                op,
                right: Box::new(right),
            };
        }
        Ok(left)
    }

    fn parse_multiplicative(&mut self) -> Result<Expression, String> {
        let mut left = self.parse_unary()?;
        loop {
            let op = match self.peek() {
                Token::Star  => BinaryOperator::Mul,
                Token::Slash => BinaryOperator::Div,
                _ => break,
            };
            self.advance();
            let right = self.parse_unary()?;
            left = Expression::BinaryOp {
                left: Box::new(left),
                op,
                right: Box::new(right),
            };
        }
        Ok(left)
    }

    fn parse_unary(&mut self) -> Result<Expression, String> {
        if self.peek() == &Token::Minus {
            self.advance();
            let expr = self.parse_unary()?; // right-associative
            Ok(Expression::UnaryOp {
                op: UnaryOperator::Neg,
                expr: Box::new(expr),
            })
        } else {
            self.parse_primary()
        }
    }

    fn parse_primary(&mut self) -> Result<Expression, String> {
        match self.peek().clone() {
            // Parenthesised expression
            Token::LParen => {
                self.advance();
                let expr = self.parse_expr()?;
                self.expect(&Token::RParen)?;
                Ok(expr)
            }

            // Literals
            Token::Integer(n) => {
                self.advance();
                Ok(Expression::Literal(Literal::Integer(n)))
            }
            Token::Float(f) => {
                self.advance();
                Ok(Expression::Literal(Literal::Float(f)))
            }
            Token::StringLit(s) => {
                self.advance();
                Ok(Expression::Literal(Literal::Str(s)))
            }
            Token::True => {
                self.advance();
                Ok(Expression::Literal(Literal::Boolean(true)))
            }
            Token::False => {
                self.advance();
                Ok(Expression::Literal(Literal::Boolean(false)))
            }
            Token::Null => {
                self.advance();
                Ok(Expression::Literal(Literal::Null))
            }

            // Identifier: column, qualified column (t.col), table wildcard (t.*), or function call
            Token::Identifier(name) => {
                self.advance();

                // Function call: name(...)
                if self.peek() == &Token::LParen {
                    self.advance(); // consume '('
                    let args = if self.peek() == &Token::RParen {
                        Vec::new()
                    } else if self.peek() == &Token::Star {
                        // COUNT(*) — consume '*', yield Wildcard arg
                        self.advance();
                        vec![Expression::Wildcard]
                    } else {
                        self.parse_expr_list()?
                    };
                    self.expect(&Token::RParen)?;
                    return Ok(Expression::Function { name, args });
                }

                // Qualified: table.column or table.*
                if self.peek() == &Token::Dot {
                    self.advance(); // consume '.'
                    return match self.advance() {
                        Token::Identifier(col) => Ok(Expression::Column {
                            table: Some(name),
                            column: col,
                        }),
                        Token::Star => {
                            // table.* — represented as Column { table, column: "*" }
                            // so parse_select_list can convert it to QualifiedWildcard
                            Ok(Expression::Column {
                                table: Some(name),
                                column: "*".to_string(),
                            })
                        }
                        tok => Err(format!("expected column name after '.', got {:?}", tok)),
                    };
                }

                Ok(Expression::Column { table: None, column: name })
            }

            tok => Err(format!("unexpected token in expression: {:?}", tok)),
        }
    }
}

// ── Tests ────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::lexer::Lexer;

    fn parse(sql: &str) -> Statement {
        let tokens = Lexer::new(sql).tokenize().expect("lex error");
        Parser::new(tokens).parse().expect("parse error")
    }

    // Helper: parse the WHERE expression from a query
    fn where_expr(sql: &str) -> Expression {
        match parse(sql) {
            Statement::Select(s) => s.where_clause.expect("no WHERE clause"),
        }
    }

    // ── 1.1  Basic structural parsing ────────────────────────────────────────

    #[test]
    fn test_select_star() {
        let stmt = parse("SELECT * FROM orders");
        let Statement::Select(s) = stmt;
        assert!(matches!(s.columns[0], SelectItem::Wildcard));
        assert_eq!(s.from.table_name, "orders");
    }

    #[test]
    fn test_select_columns_with_alias() {
        let stmt = parse("SELECT id, name AS n FROM users");
        let Statement::Select(s) = stmt;
        assert_eq!(s.columns.len(), 2);
        if let SelectItem::Expression(_, alias) = &s.columns[1] {
            assert_eq!(alias.as_deref(), Some("n"));
        } else {
            panic!("expected Expression item");
        }
    }

    #[test]
    fn test_qualified_column() {
        let stmt = parse("SELECT c.name FROM customers c WHERE c.id = 1");
        let Statement::Select(s) = stmt;
        if let SelectItem::Expression(Expression::Column { table, column }, _) = &s.columns[0] {
            assert_eq!(table.as_deref(), Some("c"));
            assert_eq!(column, "name");
        } else {
            panic!("expected qualified column");
        }
    }

    #[test]
    fn test_inner_join() {
        let stmt = parse(
            "SELECT * FROM orders o INNER JOIN customers c ON o.customer_id = c.id",
        );
        let Statement::Select(s) = stmt;
        assert_eq!(s.joins.len(), 1);
        assert_eq!(s.joins[0].join_type, JoinType::Inner);
        assert_eq!(s.joins[0].table.table_name, "customers");
    }

    #[test]
    fn test_group_by_having() {
        let stmt = parse(
            "SELECT dept, COUNT(*) FROM emp GROUP BY dept HAVING COUNT(*) > 5",
        );
        let Statement::Select(s) = stmt;
        assert_eq!(s.group_by.len(), 1);
        assert!(s.having.is_some());
    }

    #[test]
    fn test_order_by_limit() {
        let stmt = parse("SELECT id FROM t ORDER BY id DESC LIMIT 10");
        let Statement::Select(s) = stmt;
        assert_eq!(s.order_by.len(), 1);
        assert!(!s.order_by[0].ascending);
        assert_eq!(s.limit, Some(10));
    }

    #[test]
    fn test_function_call() {
        let stmt = parse("SELECT COUNT(*), AVG(price) FROM products");
        let Statement::Select(s) = stmt;
        if let SelectItem::Expression(Expression::Function { name, args }, _) = &s.columns[0] {
            assert_eq!(name, "COUNT");
            assert!(matches!(args[0], Expression::Wildcard));
        } else {
            panic!("expected function");
        }
    }

    // ── 1.2  Expression precedence ───────────────────────────────────────────

    /// `a = 1 OR b = 2 AND c = 3`  must parse as  `OR(a=1, AND(b=2, c=3))`
    #[test]
    fn test_and_binds_tighter_than_or() {
        let expr = where_expr("SELECT 1 FROM t WHERE a = 1 OR b = 2 AND c = 3");
        // Top-level must be OR
        match expr {
            Expression::BinaryOp { op: BinaryOperator::Or, left, right } => {
                // left  = a = 1
                assert!(matches!(*left, Expression::BinaryOp { op: BinaryOperator::Eq, .. }));
                // right = AND(b=2, c=3)
                assert!(matches!(*right, Expression::BinaryOp { op: BinaryOperator::And, .. }));
            }
            other => panic!("expected OR at top level, got {:?}", other),
        }
    }

    /// `a + b * c`  must parse as  `a + (b * c)`
    #[test]
    fn test_mul_binds_tighter_than_add() {
        let expr = where_expr("SELECT 1 FROM t WHERE a + b * c = 0");
        match expr {
            Expression::BinaryOp { op: BinaryOperator::Eq, left, .. } => match *left {
                Expression::BinaryOp { op: BinaryOperator::Add, right, .. } => {
                    assert!(matches!(*right, Expression::BinaryOp { op: BinaryOperator::Mul, .. }));
                }
                other => panic!("expected Add, got {:?}", other),
            },
            other => panic!("expected Eq at top, got {:?}", other),
        }
    }

    /// `(a + b) * c`  — parentheses override precedence
    #[test]
    fn test_parentheses_override_precedence() {
        let expr = where_expr("SELECT 1 FROM t WHERE (a + b) * c = 0");
        match expr {
            Expression::BinaryOp { op: BinaryOperator::Eq, left, .. } => match *left {
                Expression::BinaryOp { op: BinaryOperator::Mul, left: inner, .. } => {
                    assert!(matches!(*inner, Expression::BinaryOp { op: BinaryOperator::Add, .. }));
                }
                other => panic!("expected Mul, got {:?}", other),
            },
            other => panic!("expected Eq at top, got {:?}", other),
        }
    }

    /// `NOT a = 1 AND b = 2`  =>  `AND(NOT(a=1), b=2)` — NOT binds tighter than AND
    #[test]
    fn test_not_binds_tighter_than_and() {
        let expr = where_expr("SELECT 1 FROM t WHERE NOT a = 1 AND b = 2");
        match expr {
            Expression::BinaryOp { op: BinaryOperator::And, left, .. } => {
                assert!(matches!(*left, Expression::UnaryOp { op: UnaryOperator::Not, .. }));
            }
            other => panic!("expected AND at top, got {:?}", other),
        }
    }

    /// Unary minus: `-a + b` => `Add(Neg(a), b)`
    #[test]
    fn test_unary_minus_precedence() {
        let expr = where_expr("SELECT 1 FROM t WHERE -a + b = 0");
        match expr {
            Expression::BinaryOp { op: BinaryOperator::Eq, left, .. } => match *left {
                Expression::BinaryOp { op: BinaryOperator::Add, left: neg, .. } => {
                    assert!(matches!(*neg, Expression::UnaryOp { op: UnaryOperator::Neg, .. }));
                }
                other => panic!("expected Add, got {:?}", other),
            },
            other => panic!("expected Eq at top, got {:?}", other),
        }
    }
}

File: cpp/src/parser/parser.cpp
#include "parser.hpp"

#include <stdexcept>

namespace sql::parser {

Parser::Parser(std::vector<Token> tokens) : tokens_(std::move(tokens)) {}

const Token& Parser::peek() const {
    static const Token eof(TokenKind::Eof);
    return pos_ < tokens_.size() ? tokens_[pos_] : eof;
}

Token Parser::advance() {
    Token tok = pos_ < tokens_.size() ? tokens_[pos_] : Token(TokenKind::Eof);
    if (pos_ < tokens_.size()) ++pos_;
    return tok;
}

void Parser::expect(TokenKind expected) {
    if (peek().kind == expected) {
        advance();
        return;
    }
    throw std::runtime_error("unexpected token during parse");
}

// ── Top-level entry point ────────────────────────────────────────────────────

Statement Parser::parse() {
    if (peek().kind == TokenKind::Select) return parse_select();
    throw std::runtime_error("expected SELECT");
}

// ── SELECT statement ──────────────────────────────────────────────────────────

Statement Parser::parse_select() {
    advance(); // consume SELECT

    SelectStatement stmt;
    stmt.columns = parse_select_list();

    expect(TokenKind::From);
    stmt.from = parse_table_ref();

    // Zero or more JOINs
    while (peek().kind == TokenKind::Join || peek().kind == TokenKind::Inner ||
           peek().kind == TokenKind::Left || peek().kind == TokenKind::Right ||
           peek().kind == TokenKind::Cross) {
        stmt.joins.push_back(parse_join());
    }

    if (peek().kind == TokenKind::Where) {
        advance();
        stmt.where_clause = parse_expr();
    }

    if (peek().kind == TokenKind::Group) {
        advance();
        expect(TokenKind::By);
        stmt.group_by = parse_expr_list();
    }

    if (peek().kind == TokenKind::Having) {
        advance();
        stmt.having = parse_expr();
    }

    if (peek().kind == TokenKind::Order) {
        advance();
        expect(TokenKind::By);
        stmt.order_by = parse_order_by_list();
    }

    if (peek().kind == TokenKind::Limit) {
        advance();
        Token tok = advance();
        if (tok.kind != TokenKind::Integer) {
            throw std::runtime_error("expected integer after LIMIT");
        }
        stmt.limit = static_cast<size_t>(tok.integer());
    }

    return Statement::make_select(std::move(stmt));
}

// ── SELECT list ───────────────────────────────────────────────────────────────

std::vector<SelectItem> Parser::parse_select_list() {
    std::vector<SelectItem> items;
    for (;;) {
        if (peek().kind == TokenKind::Star) {
            advance();
            items.push_back(SelectItem::make_wildcard());
        } else {
            Expression expr = parse_expr();

            // `table.*` is emitted by parse_primary as Column{table: Some(t), column: "*"}.
            // Convert it to QualifiedWildcard here in the SELECT-list context.
            if (expr.kind == Expression::Kind::Column && expr.table.has_value() && expr.column == "*") {
                items.push_back(SelectItem::make_qualified_wildcard(*expr.table));
            } else {
                std::optional<std::string> alias;
                if (peek().kind == TokenKind::As) {
                    advance();
                    Token tok = advance();
                    if (tok.kind != TokenKind::Identifier) {
                        throw std::runtime_error("expected alias name after AS");
                    }
                    alias = tok.text();
                } else if (peek().kind == TokenKind::Identifier) {
                    // Implicit alias: `expr name`
                    alias = advance().text();
                }
                items.push_back(SelectItem::make_expression(std::move(expr), std::move(alias)));
            }
        }

        if (peek().kind == TokenKind::Comma) {
            advance();
        } else {
            break;
        }
    }
    return items;
}

// ── Table reference ───────────────────────────────────────────────────────────

TableRef Parser::parse_table_ref() {
    Token name_tok = advance();
    if (name_tok.kind != TokenKind::Identifier) {
        throw std::runtime_error("expected table name");
    }

    TableRef ref;
    ref.table_name = name_tok.text();

    if (peek().kind == TokenKind::As) {
        advance();
        Token tok = advance();
        if (tok.kind != TokenKind::Identifier) throw std::runtime_error("expected alias after AS");
        ref.alias = tok.text();
    } else if (peek().kind == TokenKind::Identifier) {
        ref.alias = advance().text();
    }

    return ref;
}

// ── JOIN clause ────────────────────────────────────────────────────────────────

Join Parser::parse_join() {
    JoinType join_type;
    switch (peek().kind) {
        case TokenKind::Join:
            advance();
            join_type = JoinType::Inner;
            break;
        case TokenKind::Inner:
            advance();
            expect(TokenKind::Join);
            join_type = JoinType::Inner;
            break;
        case TokenKind::Left:
            advance();
            expect(TokenKind::Join);
            join_type = JoinType::Left;
            break;
        case TokenKind::Right:
            advance();
            expect(TokenKind::Join);
            join_type = JoinType::Right;
            break;
        case TokenKind::Cross:
            advance();
            expect(TokenKind::Join);
            join_type = JoinType::Cross;
            break;
        default:
            throw std::runtime_error("expected JOIN keyword");
    }

    Join join;
    join.join_type = join_type;
    join.table = parse_table_ref();
    expect(TokenKind::On);
    join.condition = parse_expr();
    return join;
}

// ── ORDER BY list ──────────────────────────────────────────────────────────────

std::vector<OrderByItem> Parser::parse_order_by_list() {
    std::vector<OrderByItem> items;
    for (;;) {
        Expression expr = parse_expr();
        bool ascending = true;
        if (peek().kind == TokenKind::Asc) {
            advance();
            ascending = true;
        } else if (peek().kind == TokenKind::Desc) {
            advance();
            ascending = false;
        }
        items.push_back(OrderByItem{std::move(expr), ascending});

        if (peek().kind == TokenKind::Comma) {
            advance();
        } else {
            break;
        }
    }
    return items;
}

// ── Comma-separated expression list ───────────────────────────────────────────

std::vector<Expression> Parser::parse_expr_list() {
    std::vector<Expression> exprs;
    for (;;) {
        exprs.push_back(parse_expr());
        if (peek().kind == TokenKind::Comma) {
            advance();
        } else {
            break;
        }
    }
    return exprs;
}

// ── Expression parsing (recursive descent by precedence) ──────────────────────

Expression Parser::parse_expr() { return parse_or(); }

Expression Parser::parse_or() {
    Expression left = parse_and();
    while (peek().kind == TokenKind::Or) {
        advance();
        Expression right = parse_and();
        left = Expression::make_binary_op(std::move(left), BinaryOperator::Or, std::move(right));
    }
    return left;
}

Expression Parser::parse_and() {
    Expression left = parse_not();
    while (peek().kind == TokenKind::And) {
        advance();
        Expression right = parse_not();
        left = Expression::make_binary_op(std::move(left), BinaryOperator::And, std::move(right));
    }
    return left;
}

Expression Parser::parse_not() {
    if (peek().kind == TokenKind::Not) {
        advance();
        Expression expr = parse_not(); // right-associative: NOT NOT a => Not(Not(a))
        return Expression::make_unary_op(UnaryOperator::Not, std::move(expr));
    }
    return parse_comparison();
}

Expression Parser::parse_comparison() {
    Expression left = parse_additive();
    BinaryOperator op;
    switch (peek().kind) {
        case TokenKind::Eq:  op = BinaryOperator::Eq;  break;
        case TokenKind::Neq: op = BinaryOperator::Neq; break;
        case TokenKind::Lt:  op = BinaryOperator::Lt;  break;
        case TokenKind::Gt:  op = BinaryOperator::Gt;  break;
        case TokenKind::Lte: op = BinaryOperator::Lte; break;
        case TokenKind::Gte: op = BinaryOperator::Gte; break;
        default: return left;
    }
    advance();
    Expression right = parse_additive();
    return Expression::make_binary_op(std::move(left), op, std::move(right));
}

Expression Parser::parse_additive() {
    Expression left = parse_multiplicative();
    for (;;) {
        BinaryOperator op;
        if (peek().kind == TokenKind::Plus) op = BinaryOperator::Add;
        else if (peek().kind == TokenKind::Minus) op = BinaryOperator::Sub;
        else break;
        advance();
        Expression right = parse_multiplicative();
        left = Expression::make_binary_op(std::move(left), op, std::move(right));
    }
    return left;
}

Expression Parser::parse_multiplicative() {
    Expression left = parse_unary();
    for (;;) {
        BinaryOperator op;
        if (peek().kind == TokenKind::Star) op = BinaryOperator::Mul;
        else if (peek().kind == TokenKind::Slash) op = BinaryOperator::Div;
        else break;
        advance();
        Expression right = parse_unary();
        left = Expression::make_binary_op(std::move(left), op, std::move(right));
    }
    return left;
}

Expression Parser::parse_unary() {
    if (peek().kind == TokenKind::Minus) {
        advance();
        Expression expr = parse_unary(); // right-associative
        return Expression::make_unary_op(UnaryOperator::Neg, std::move(expr));
    }
    return parse_primary();
}

Expression Parser::parse_primary() {
    const Token tok = peek();

    switch (tok.kind) {
        case TokenKind::LParen: {
            advance();
            Expression expr = parse_expr();
            expect(TokenKind::RParen);
            return expr;
        }
        case TokenKind::Integer:
            advance();
            return Expression::make_literal(Literal::integer(tok.integer()));
        case TokenKind::Float:
            advance();
            return Expression::make_literal(Literal::floating(tok.floating()));
        case TokenKind::StringLit:
            advance();
            return Expression::make_literal(Literal::str(tok.text()));
        case TokenKind::True:
            advance();
            return Expression::make_literal(Literal::boolean(true));
        case TokenKind::False:
            advance();
            return Expression::make_literal(Literal::boolean(false));
        case TokenKind::Null:
            advance();
            return Expression::make_literal(Literal::null());

        case TokenKind::Identifier: {
            std::string name = tok.text();
            advance();

            // Function call: name(...)
            if (peek().kind == TokenKind::LParen) {
                advance(); // consume '('
                std::vector<Expression> args;
                if (peek().kind == TokenKind::RParen) {
                    // no args
                } else if (peek().kind == TokenKind::Star) {
                    // COUNT(*) — consume '*', yield Wildcard arg
                    advance();
                    args.push_back(Expression::make_wildcard());
                } else {
                    args = parse_expr_list();
                }
                expect(TokenKind::RParen);
                return Expression::make_function(std::move(name), std::move(args));
            }

            // Qualified: table.column or table.*
            if (peek().kind == TokenKind::Dot) {
                advance(); // consume '.'
                Token next = advance();
                if (next.kind == TokenKind::Identifier) {
                    return Expression::make_column(std::move(name), next.text());
                }
                if (next.kind == TokenKind::Star) {
                    // table.* — represented as Column{table, column: "*"} so
                    // parse_select_list can convert it to QualifiedWildcard.
                    return Expression::make_column(std::move(name), "*");
                }
                throw std::runtime_error("expected column name after '.'");
            }

            return Expression::make_column(std::nullopt, std::move(name));
        }

        default:
            throw std::runtime_error("unexpected token in expression");
    }
}

} // namespace sql::parser
